# Homework 2 for "Transcriptomics and Metagenomics" course

#### [Assignmnent description](https://docs.google.com/document/d/15QH3CC-UtinXM76fyToKoLiQD0YLnpK-TUfplt6mEks/edit)

## 0. Downloading necessary data

#### 0.1. E. coli reference genome

In [ ]:
# assuming working directory is ~/raw_data

# !wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz
# !wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz

#### 0.2. E. coli resistant to Amp (raw Illumina sequencing reads)

In [ ]:
# https://figshare.com/articles/dataset/amp_res_2_fastq_zip/10006541/3?file=23769692

## 1. Manual inspection of reads

### 1.1. Checking the contents

In [14]:
# printing first 20 lines of file

!head -20 /home/fpolyakov/trans_meta/homework2/raw_data/amp_res_1.fastq

@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101
GGTTGCAGATTCGCAGTGTCGCTGTTCCAGCGCATCACATCTTTGATGTTCACGCCGTGGCGTTTAGCAATGCTTGAAAGCGAATCGCCTTTGCCCACACG
+
@?:=:;DBFADH;CAECEE@@E:FFHGAE4?C?DE<BFGEC>?>FHE4BFFIIFHIBABEECA83;>>@>@CCCDC9@@CC08<@?@BB@9:CC#######
@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101
GTATGAGGTTTTGCTGCATTCTCTGNGCGAATATTAACTCCNTNNNNNTTATAGTTCAAAGCAAGTACCTGTCTCTTATACACATCTCCGAGCCCACGAGC
+
@@<?=D?D==?<AFGDF+AIHEACH#22<:?E8??:9??GG#0#####000;CF=C)4.==CA@@@)=7?C7?E37;3@>;;(.;>AB#############
@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101
GCTTCTCTTAACTGAGGTCACCATCATGCCGTTAAGTCCCTACCTCTCTTTTGCCGGTAACTGTTCCGCCGCGATTGCCTTTTATCTGTCTCTTATACACC
+
??<DBD;4C2=<BB>:AC;<CF<CE@FE9@E1C@891CD*9:?:3D@DD4?D<DD:0;@A=AEIDDA##################################
@SRR1363257.78 GWZHISEQ01:153:C1W31ACXX:5:1101:5178:2440 length=101
GCATAAGGACGATCGCTCCAGAGTAAAATAAATACGCGCATGTGATACTCACAATACCAATGGTGAAGTTACGGGACTTAAACAAACTGAGATCAAGAATC
+
CCCF

### 1.2. Determining the number of reads in files

In [20]:
# for amp_res_1.fastq

!echo $(cat /home/fpolyakov/trans_meta/homework2/raw_data/amp_res_1.fastq|wc -l)/4|bc

455876


In [21]:
# for amp_res_2.fastq

!echo $(cat /home/fpolyakov/trans_meta/homework2/raw_data/amp_res_2.fastq|wc -l)/4|bc

455876


## 2. Running FastQC

In [22]:
%%bash

/home/fpolyakov/trans_meta/homework2/soft/FastQC/fastqc \
-o /home/fpolyakov/trans_meta/homework2/results/ \
/home/fpolyakov/trans_meta/homework2/raw_data/amp_res_1.fastq \
/home/fpolyakov/trans_meta/homework2/raw_data/amp_res_2.fastq

null
null


Started analysis of amp_res_1.fastq
Approx 5% complete for amp_res_1.fastq
Approx 10% complete for amp_res_1.fastq
Approx 15% complete for amp_res_1.fastq
Approx 20% complete for amp_res_1.fastq
Approx 25% complete for amp_res_1.fastq
Approx 30% complete for amp_res_1.fastq
Approx 35% complete for amp_res_1.fastq
Approx 40% complete for amp_res_1.fastq
Approx 45% complete for amp_res_1.fastq
Approx 50% complete for amp_res_1.fastq
Approx 55% complete for amp_res_1.fastq
Approx 60% complete for amp_res_1.fastq
Approx 65% complete for amp_res_1.fastq
Approx 70% complete for amp_res_1.fastq
Approx 75% complete for amp_res_1.fastq
Approx 80% complete for amp_res_1.fastq
Approx 85% complete for amp_res_1.fastq
Approx 90% complete for amp_res_1.fastq
Approx 95% complete for amp_res_1.fastq


Analysis complete for amp_res_1.fastq


Started analysis of amp_res_2.fastq
Approx 5% complete for amp_res_2.fastq
Approx 10% complete for amp_res_2.fastq
Approx 15% complete for amp_res_2.fastq
Approx 20% complete for amp_res_2.fastq
Approx 25% complete for amp_res_2.fastq
Approx 30% complete for amp_res_2.fastq
Approx 35% complete for amp_res_2.fastq
Approx 40% complete for amp_res_2.fastq
Approx 45% complete for amp_res_2.fastq
Approx 50% complete for amp_res_2.fastq
Approx 55% complete for amp_res_2.fastq
Approx 60% complete for amp_res_2.fastq
Approx 65% complete for amp_res_2.fastq
Approx 70% complete for amp_res_2.fastq
Approx 75% complete for amp_res_2.fastq
Approx 80% complete for amp_res_2.fastq
Approx 85% complete for amp_res_2.fastq
Approx 90% complete for amp_res_2.fastq
Approx 95% complete for amp_res_2.fastq


Analysis complete for amp_res_2.fastq


## 3. Running BWA

### 3.1. Indexing the reference file

In [25]:
!/home/fpolyakov/trans_meta/homework2/soft/bwa/bwa index GCF_000005845.2_ASM584v2_genomic.fna

[bwa_index] Pack FASTA... 0.05 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 1.00 seconds elapse.
[bwa_index] Update BWT... 0.02 sec
[bwa_index] Pack forward-only FASTA... 0.02 sec
[bwa_index] Construct SA from BWT and Occ... 0.41 sec
[main] Version: 0.7.17-r1198-dirty
[main] CMD: /home/fpolyakov/trans_meta/homework2/soft/bwa/bwa index GCF_000005845.2_ASM584v2_genomic.fna
[main] Real time: 1.676 sec; CPU: 1.512 sec


In [27]:
!ls

amp_res_1.fastq
amp_res_2.fastq
GCF_000005845.2_ASM584v2_genomic.fna
GCF_000005845.2_ASM584v2_genomic.fna.amb
GCF_000005845.2_ASM584v2_genomic.fna.ann
GCF_000005845.2_ASM584v2_genomic.fna.bwt
GCF_000005845.2_ASM584v2_genomic.fna.pac
GCF_000005845.2_ASM584v2_genomic.fna.sa
GCF_000005845.2_ASM584v2_genomic.gff


### 3.2. Aligning reads

In [29]:
%%bash

/home/fpolyakov/trans_meta/homework2/soft/bwa/bwa mem \
/home/fpolyakov/trans_meta/homework2/raw_data/GCF_000005845.2_ASM584v2_genomic.fna \
/home/fpolyakov/trans_meta/homework2/raw_data/amp_res_1.fastq \
/home/fpolyakov/trans_meta/homework2/raw_data/amp_res_2.fastq \
> alignment.sam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 99010 sequences (10000010 bp)...
[M::process] read 99010 sequences (10000010 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 47869, 0, 15)
[M::mem_pestat] skip orientation FF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (146, 185, 231)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 401)
[M::mem_pestat] mean and std.dev: (190.00, 63.93)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 486)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] analyzing insert size distribution for orientation RR...
[M::mem_pestat] (25, 50, 75) percentile: (120, 158, 1540)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4380)
[M::mem_pestat] mean and std.dev: (568.27, 721.54)
[M::mem_pestat] low and high boundaries for pr

### 3.3. Compressing SAM file

In [31]:
# converting sam to bam

!/home/fpolyakov/trans_meta/homework2/soft/samtools-1.18/samtools view -S -b alignment.sam > alignment.bam 

In [32]:
# getting some stats

!/home/fpolyakov/trans_meta/homework2/soft/samtools-1.18/samtools flagstat alignment.bam

912095 + 0 in total (QC-passed reads + QC-failed reads)
911752 + 0 primary
0 + 0 secondary
343 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
910940 + 0 mapped (99.87% : N/A)
910597 + 0 primary mapped (99.87% : N/A)
911752 + 0 paired in sequencing
455876 + 0 read1
455876 + 0 read2
907476 + 0 properly paired (99.53% : N/A)
909488 + 0 with itself and mate mapped
1109 + 0 singletons (0.12% : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


#### We can see that 99.87% of reads successfully mapped

### 3.4. Sorting and indexing BAM file

#### 3.4.1. Sorting bam file by sequence coordinate on reference


In [33]:
!/home/fpolyakov/trans_meta/homework2/soft/samtools-1.18/samtools sort alignment.bam -o alignment_sorted.bam

#### 3.4.2. Indexing BAM file for faster search

In [34]:
!/home/fpolyakov/trans_meta/homework2/soft/samtools-1.18/samtools index alignment_sorted.bam

## 4. Variant calling

### 4.1. Creating intermediate file

In [36]:
!/home/fpolyakov/trans_meta/homework2/soft/samtools-1.18/samtools mpileup -f /home/fpolyakov/trans_meta/homework2/raw_data/GCF_000005845.2_ASM584v2_genomic.fna alignment_sorted.bam >  my.mpileup

[mpileup] 1 samples in 1 input files


In [37]:
os.chdir('/home/fpolyakov/trans_meta/homework2/soft/varscan/')

### 4.2. VarScan analysis

In [38]:
!java -jar VarScan.v2.4.0.jar mpileup2snp /home/fpolyakov/trans_meta/homework2/raw_data/my.mpileup --min-var-freq 0.05 --variants --output-vcf 1 > VarScan_results.vcf


Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.05
Min avg qual:	15
P-value thresh:	0.01
Reading input from /home/fpolyakov/trans_meta/homework2/raw_data/my.mpileup
4641524 bases in pileup file
9 variant positions (6 SNP, 3 indel)
0 were failed by the strand-filter
6 variant positions reported (6 SNP, 0 indel)


## 5. Post-analysis

### The rest work was done manually using terminal